### PR TITLE
finish homework

### PR DESCRIPTION
**Pull request by zero1608**

### CODE REVIEW

File: src/main/java/com/epam/processor/DataProcessor.java [hunk 1]
```diff
@@ -1,123 +1,240 @@
...
+	 * @return
+	 */
+	public RoadAccident getAccidentByIndex7(String index) {
+		for (RoadAccident roadAccident : roadAccidentList) {
+			if (roadAccident.getAccidentId().equals(index)) {
+				return roadAccident;
+			}
+		}
+		return null;
+	}
```
Comment (anil-kapoor) [inline src/main/java/com/epam/processor/DataProcessor.java:48]: @zero1608 
This is good.

File: src/main/java/com/epam/processor/DataProcessor.java [hunk 2]
```diff
@@ -1,123 +1,240 @@
...
+		for (RoadAccident roadAccident : roadAccidentList) {
+			if (roadAccident.getLongitude() >= minLongitude
+					&& roadAccident.getLongitude() <= maxLongitude
+					&& roadAccident.getLatitude() >= minLatitude
+					&& roadAccident.getLatitude() <= maxLatitude) {
+				list.add(roadAccident);
+			}
+		}
+		return list;
+	}
```
Comment (anil-kapoor) [inline src/main/java/com/epam/processor/DataProcessor.java:71]: @zero1608 
This is good. Try to use some meaningful names for variables. On place of list you could use accidentList.

File: src/main/java/com/epam/processor/DataProcessor.java [hunk 3]
```diff
@@ -1,123 +1,240 @@
...
+			} else if (roadAccident.getRoadSurfaceConditions().equals("5")
+					|| roadAccident.getRoadSurfaceConditions()
+							.equalsIgnoreCase("dry")) {
+				dry++;
+			}
+		}
+		map.put("Dry", dry);
+		map.put("Wet", wet);
+		return map;
+	}
```
Comment (anil-kapoor) [inline src/main/java/com/epam/processor/DataProcessor.java:91]: @zero1608 
2 & 5 are count for dry & wet. They are not conditions. And that is just example.
Please try to implement it in a generic way so any type of condition can come in output.

File: src/main/java/com/epam/processor/DataProcessor.java [hunk 4]
```diff
@@ -1,123 +1,240 @@
...
+				countList.add(1, size);
+			} else if (size > countList.get(2)) {
+				list.add(2, key);
+				countList.add(2, size);
+			}
+			if (list.size() == 4) {
+				list.remove(3);
+			}
+		}
+		return list;
```
Comment (anil-kapoor) [inline src/main/java/com/epam/processor/DataProcessor.java]: @zero1608 
Can you explain how are you achieving the output from this???
Seems very complex to me.

File: src/main/java/com/epam/processor/DataProcessor.java [hunk 5]
```diff
@@ -1,123 +1,240 @@
...
+	 * accident ids ex: authority1 -> id1, id2, id3 authority2 -> id4, id5
+	 * 
+	 * @return
+	 */
+	public Multimap<String, String> getAccidentIdsGroupedByAuthority7() {
+		Multimap<String, String> multiMap = HashMultimap.create();
+		long count = 0L;
+		for (RoadAccident roadAccident : roadAccidentList) {
+			multiMap.put(roadAccident.getDistrictAuthority(),
+					roadAccident.getAccidentId() + count++);
```
Comment (anil-kapoor) [inline src/main/java/com/epam/processor/DataProcessor.java:142]: @zero1608 
This is not the expected behaviour, id1, id2 is just for examples. Those ids can be abc, a4, 789 anything.
Please correct.

File: src/main/java/com/epam/processor/DataProcessor.java [hunk 6]
```diff
@@ -1,123 +1,240 @@
...
+	}
+
+	// Now let's do same tasks but now with streaming api
+
+	public RoadAccident getAccidentByIndex(String index) {
+		RoadAccident roadAccident = roadAccidentList.stream()
+				.filter(ra -> ra.getAccidentId().equals(index)).findFirst()
+				.orElse(null);
+		return roadAccident;
+	}
```
Comment (anil-kapoor) [inline src/main/java/com/epam/processor/DataProcessor.java:154]: @zero1608 
This is good. :+1: 
just use meaningful names in variables.

File: src/main/java/com/epam/processor/DataProcessor.java [hunk 7]
```diff
@@ -1,123 +1,240 @@
...
+	 */
+	public Collection<RoadAccident> getAccidentsByLocation(float minLongitude,
+			float maxLongitude, float minLatitude, float maxLatitude) {
+		return roadAccidentList
+				.stream()
+				.filter(roadAccident -> roadAccident.getLongitude() >= minLongitude
+						&& roadAccident.getLongitude() <= maxLongitude)
+				.filter(roadAccident -> roadAccident.getLatitude() >= minLatitude
+						&& roadAccident.getLatitude() <= maxLatitude)
+				.collect(Collectors.toList());
```
Comment (anil-kapoor) [inline src/main/java/com/epam/processor/DataProcessor.java:173]: @zero1608 
:+1: 
for Collectors you can import it statically like import static java.util.stream.Collectors.*;

File: src/main/java/com/epam/processor/DataProcessor.java [hunk 8]
```diff
@@ -1,123 +1,240 @@
...
+						Collectors.groupingBy(item -> item,
+								Collectors.counting()));
+
+		return map
+				.entrySet()
+				.stream()
+				.sorted((val1, val2) -> val2.getValue().compareTo(
+						val1.getValue())).map(Map.Entry::getKey).limit(3)
+				.collect(Collectors.toList());
+	}
```
Comment (anil-kapoor) [inline src/main/java/com/epam/processor/DataProcessor.java:196]: @zero1608 
:+1:

File: src/main/java/com/epam/processor/DataProcessor.java [hunk 9]
```diff
@@ -1,123 +1,240 @@
...
+		long wet = roadAccidentList
+				.stream()
+				.filter(ra -> ra.getRoadSurfaceConditions().equals("2")
+						|| ra.getRoadSurfaceConditions()
+								.equalsIgnoreCase("wet")).count();
+
+		map.put("Dry", dry);
+		map.put("Wet", wet);
+		return map;
+	}
```
Comment (anil-kapoor) [inline src/main/java/com/epam/processor/DataProcessor.java:210]: @zero1608 
Your understanding is not correct on this. Please read again & change the code accordingly.

File: src/main/java/com/epam/processor/DataProcessor.java [hunk 10]
```diff
@@ -1,123 +1,240 @@
...
+	 * @return
+	 */
+	public Map<String, List<String>> getAccidentIdsGroupedByAuthority() {
+		return roadAccidentList.stream().collect(
+				Collectors.groupingBy(RoadAccident::getDistrictAuthority,
+						Collectors.mapping(RoadAccident::getAccidentId,
+								Collectors.toList())));
+	}
+
+}
```
Comment (anil-kapoor) [inline src/main/java/com/epam/processor/DataProcessor.java:225]: @zero1608 
:+1:  perfect.

**Comment by bug-hunterx** [comment]
Closing PR - homework accepted

**Pull request closed without merging**